### PR TITLE
Limit `bit_view`s to 8-bit registers to avoid byte-order assumption

### LINKED ISSUE #1993: Test case bit_view.size_safety fails on s390x architecture
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

_No response_

### What Operating System are you using?

Other

### If Other OS, please describe

Linux s390x

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

I am trying to update DOSBox Staging in Fedora to 0.79.1.
The build succeeds on all other Fedora architectures except s390x.
On s390x, there is one failing test as follows:

```
[ RUN      ] bit_view.size_safety
../tests/bit_view_tests.cpp:310: Failure
Value of: r8.first_8 == 0b000'1111
  Actual: false
Expected: true
[  FAILED  ] bit_view.size_safety (0 ms)
```

Investigating further, `r9.first_8` actually has the value that the test sets for `last_8`.
Since I know that using `union` and bitwise operations (like `bit_view` does) can lead to endianess-dependent code,
my guess is that this is related to s390x being big endian.

Testing is somewhat difficult, due to lack of hardware.
However, I have a setup where I can build and run code in Fedora's infra.

Admittedly, it seems unlikely that anybody will want to run DOSBox Staging on a mainframe,
so I could just not build for s390x,
on the basis that the test suite does not pass on that arch.
However, seeing that the codebase otherwise has support for big endian through `WORDS_BIGENDIAN`,
perhaps this is a valid bug that deserves to be fixed?

### Steps to reproduce the behaviour.

Explain how to reproduce
Build and run the test suite on s390x architecture.
For example, [dosbox-staging-build][copr] in Copr.
Check chroot `fedora-rawhide-s390x`'s logs at `builder-live.log` for gtest output.

[copr]: https://copr.fedorainfracloud.org/coprs/oturpe/dosbox-staging-build/build/4882237/

### Your configuration

```ini
n/a
```

### Provide a Log

```text
Build and test logs:
https://download.copr.fedorainfracloud.org/results/oturpe/dosbox-staging-build/fedora-rawhide-s390x/04882237-dosbox-staging/builder-live.log.gz
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by kcgen**
This PR constraints `bit_view`s to 8-bits, and will generate compile-time assert errors when managing larger types. This is because the byte order in a multi-byte union is host-dependent, and even though we know if a host is big or little endian, the data itself might be written from an always-little-endian space, such as emulated DOS memory, and therefore might not need to be flipped (even on a big-endian system).

So it moves the byte-order decision outside, to the programmer.

@FeralChild64, this touches the mouse draw cursor 16-bit result type, where you previous used two bit_views to access the low and high parts.  I followed the function calls that populate `&result` and saw the value is eventually written in host/native format, so I believe the previous code would have had the low and high parts swapped when running on a big-endian host.

This replaces those two views with `read_low_byte(..)` and `read_high_byte(..)`, which operate on the value itself instead of assuming a  byte order. I added unit tests for these two calls, and they pass on little and big endian systems.

Thanks to @oturpe for flagging this on RedHat's IBM s390x mainframe.

Fixes #1993

### CODE REVIEW

**Comment by kcgen** [comment]
@oturpe , curious if you can run this branch: [kc/constrain-bit_view-1](https://github.com/dosbox-staging/dosbox-staging/tree/kc/constrain-bit_view-1) on your big-endian system.

The bit_view testcase we're specifically interested is in:

```
[ RUN      ] bit_view.multibyte
[       OK ] bit_view.multibyte (0 ms)
```

File: src/ints/mouse_dos_driver.cpp [hunk 1]
```diff
@@ -292,28 +293,27 @@ static void draw_cursor_text()
...
+		             state.background.pos_y,
+		             page,
+		             &result); // result is in native/host-endian format
+		state.background.data[0] = read_low_byte(result);
+		state.background.data[1] = read_high_byte(result);
+		state.background.enabled = true;
+
+    	// Write Cursor
+		result &= state.text_and_mask;
+        result ^= state.text_xor_mask;
```
Comment (FeralChild64) [inline src/ints/mouse_dos_driver.cpp:308]: Flaky formatting (lines 306 and 308 should be indented more; tabs. s spaces problem?)

**Comment by FeralChild64** [comment]
If I’m not mistaken, the bit_view in the cursor drawing code was introduced due to a reviev remark. Nevertheless - looks good, save for some formatting.

File: src/ints/mouse_dos_driver.cpp (hunk 1 above)
Comment (kcgen) [reply src/ints/mouse_dos_driver.cpp:308]: Oh - good catch; I noticed `clang-format` wasn't doing a good job with it, but turns out most of the file has issues (spaces at the start of lines; maybe a historic problem).  

Will get this fixed up locally.

**Comment by kcgen** [comment]
> If I’m not mistaken, the bit_view in the cursor drawing code was introduced due to a review remark.

Indeed (that was before the concern of multi-byte union types being byte-swapped on big-endian systems).  This takes care of it :+1:

File: include/bit_view.h [hunk 2]
```diff
@@ -47,30 +50,49 @@ This gives us the benefits of bitfields without their specification downsides:
...
+  stereo_pan(reg.left_pan, reg.right_pan)
+
+8-bit limitation
+~~~~~~~~~~~~~~~~
+
+bit_views are endian-safe as they're limited wrapping 8-bit registers,
+which are not affected by the byte-ordering of larger multi-byte type.
+(compile-type assertions guarantee this as well).
+
+To use bit_views with larger types or registers, the type shold be
```
Comment (oturpe) [inline include/bit_view.h:86]: Typo
```suggestion
To use bit_views with larger types or registers, the type should be
```

File: include/byteorder.h [hunk 3]
```diff
@@ -174,4 +174,20 @@ constexpr uint8_t le_to_host(const uint8_t x)
 	return host_to_le(x);
 }
 
+// Helper functions to read the low and high bytes from a native 16-bit value
+// regardless of machine byte-order or bit-signficant-order.
```
Comment (oturpe) [inline include/byteorder.h:178]: Typo
```suggestion
// regardless of machine byte-order or bit-significance-order.
```

**Comment by oturpe** [comment]
> @oturpe , curious if you can run this branch: [kc/constrain-bit_view-1](https://github.com/dosbox-staging/dosbox-staging/tree/kc/constrain-bit_view-1) on your big-endian system.

I ran it, all the tests pass now. Thank you!

**Review by FeralChild64 — APPROVED** [review]

**Comment by kcgen** [comment]
Thanks for the review, @FeralChild64 :rocket: 
Thanks for testing it, @oturpe !

**Pull request merged**